### PR TITLE
Task-execution

### PR DESCRIPTION
**Pull request by Frosendroska**

### CODE REVIEW

File: nnbuilder-api/src/main/proto/tasksqueue.proto [hunk 1]
```diff
@@ -38,6 +39,6 @@ message GetInformationRequest {
 }
 
 message GetInformationResponse {
-  int64 timeDeltaSeconds = 1;
+  int64 timeSpentOnLearningSec = 1;
```
Comment (iskhakovt) [inline nnbuilder-api/src/main/proto/tasksqueue.proto]: -> taskProcessingTimeSeconds ?

File: nnbuilder/build.gradle [hunk 2]
```diff
@@ -39,6 +39,10 @@ dependencies {
 
   implementation "io.grpc:grpc-stub:${grpcVersion}"
   implementation "io.grpc:grpc-kotlin-stub:${grpcKotlinVersion}"
+
+  // https://mvnrepository.com/artifact/commons-codec/commons-codec
+  implementation group: 'commons-codec', name: 'commons-codec', version: '1.15'
```
Comment (iskhakovt) [inline nnbuilder/build.gradle]: nit: -> `implementation 'commons-codec:commons-codec:1.15`
also, no need for the comment

File: nnbuilder/src/main/kotlin/org/hse/nnbuilder/utils/HashUtil.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import java.nio.file.StandardCopyOption;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.hse.nnbuilder.exception.HashError;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Helps to get a hash of file and hash of a String using sha1Hex
+ */
+public class HashUtil {
+    public static String getFileHash(@NotNull Path filePath) throws HashError {
```
Comment (iskhakovt) [inline nnbuilder/src/main/kotlin/org/hse/nnbuilder/utils/HashUtil.java]: What's the advantage of declaring the throw here?
It should be sufficient to throw RuntimeError (or a derivative) without declaring `throws`.

File: nnbuilder/src/main/kotlin/org/hse/nnbuilder/utils/HashUtil.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.hse.nnbuilder.exception.HashError;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Helps to get a hash of file and hash of a String using sha1Hex
+ */
+public class HashUtil {
```
Comment (iskhakovt) [inline nnbuilder/src/main/kotlin/org/hse/nnbuilder/utils/HashUtil.java]: An util class must be final and have a private constructor.

File: nnbuilder/src/main/kotlin/org/hse/nnbuilder/utils/HashUtil.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.hse.nnbuilder.exception.HashError;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Helps to get a hash of file and hash of a String using sha1Hex
```
Comment (iskhakovt) [inline nnbuilder/src/main/kotlin/org/hse/nnbuilder/utils/HashUtil.java]: "Helps to" is wordy, it doesn't "help", it does the hashing :)

File: nnbuilder/src/main/kotlin/org/hse/nnbuilder/utils/HashUtil.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    public static String getFileHash(@NotNull Path filePath) throws HashError {
+        try (InputStream is = Files.newInputStream(filePath)) {
+            return DigestUtils.sha1Hex(is);
+        } catch (IOException e) {
+            throw new HashError("Impossible to hash " + filePath.getFileName() + ".", e);
+        }
+    }
+
+    public static String getStringHash(@NotNull String commitInfo) {
+        return DigestUtils.sha1Hex(commitInfo);
```
Comment (iskhakovt) [inline nnbuilder/src/main/kotlin/org/hse/nnbuilder/utils/HashUtil.java:21]: Use sha256 or sha512, sha1 is old.

File: nnbuilder/src/main/resources/application-test.properties [hunk 7]
```diff
@@ -3,21 +3,21 @@ server.port=8000
...
 #spring.jpa.generate-ddl=true
 
+#Your current local settings for db (maybe different)
+
+spring.datasource.url=jdbc:postgresql://localhost:5432/nnbuilder
+spring.datasource.jdbc-url=${spring.datasource.url}
+spring.datasource.driver-class-name=org.postgresql.Driver
+spring.datasource.platform=postgres
+spring.datasource.username=nnbuilder
+spring.datasource.password=nnbuilder
```
Comment (iskhakovt) [inline nnbuilder/src/main/resources/application-test.properties]: This is likely to break the ci tests, can we populate these from `.env`?

File: nnbuilder/src/test/kotlin/org/hse/nnbuilder/queue/TaskExecutorTest.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+@SpringBootTest
+@ActiveProfiles("test")
+public class TaskExecutorTest {
+
+    @Autowired
+    TaskExecutor taskExecutor;
+
+    @Test
+    void callProcessTaskTest() {
+        taskExecutor.processTask();
```
Comment (iskhakovt) [inline nnbuilder/src/test/kotlin/org/hse/nnbuilder/queue/TaskExecutorTest.java]: So what does it test? :)

It should be reasonable to test the logic with mocks.

File: nnbuilder/src/test/kotlin/org/hse/nnbuilder/queue/TasksQueueRepositoryTest.java [hunk 9]
```diff
@@ -26,7 +26,7 @@ public class TasksQueueRepositoryTest {
     @Autowired
     private TaskQueuedStorage taskQueuedStorage;
 
-    final Long defaultEpochAmount = 100L;
+    final static Long DEFAULT_EPOCH_AMOUNT = 100L;
```
Comment (iskhakovt) [inline nnbuilder/src/test/kotlin/org/hse/nnbuilder/queue/TasksQueueRepositoryTest.java]: amount -> number ?

File: nnbuilder/src/main/kotlin/org/hse/nnbuilder/queue/TaskExecutor.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+                return;
+            }
+            Process exec = Runtime.getRuntime().exec(
+                    new String[] {
+                            pythonPackageInstaller.getPackageExecutablePath(), task.getTaskId().toString()
+                    }
+            );
+
+            // Wait till the end of the learning process
+            exec.waitFor();
```
Comment (iskhakovt) [inline nnbuilder/src/main/kotlin/org/hse/nnbuilder/queue/TaskExecutor.java:33]: Report failed based on return code ?

File: nnbuilder/src/main/kotlin/org/hse/nnbuilder/queue/TaskExecutor.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+            //
+
+            setFinishTaskTime(task);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Transactional
+    public void setFinishTaskTime(TaskQueued task) {
```
Comment (iskhakovt) [inline nnbuilder/src/main/kotlin/org/hse/nnbuilder/queue/TaskExecutor.java:43]: Make the function name reflect all side effects

File: nnbuilder/src/main/kotlin/org/hse/nnbuilder/queue/TaskExecutor.java (hunk 10 above)
Comment (Frosendroska) [reply nnbuilder/src/main/kotlin/org/hse/nnbuilder/queue/TaskExecutor.java:33]: base on whether Python set taskType as Done or no)))

File: nnbuilder/src/main/resources/application-test.properties (hunk 7 above)
Comment (Frosendroska) [reply nnbuilder/src/main/resources/application-test.properties]: ?

**Review by iskhakovt — APPROVED** [review]

**Pull request merged**